### PR TITLE
Add per route Middleware 

### PR DESCRIPTION
**Pull request by ryanhall07**
This allows each route to configure the Middleware that runs prior to
the actual handler. This gives each route the ability to
enable/disable/configure common Middleware functions.

This replaces the global middleware and forces each route to define the
set of middleware (or inherit the default). The ordering of middleware
functions is extremely important and potentially error prone, so best
not to have 2 levels of middleware functions.

Additionally this installs the source.Middleware function for all query
endpoints. This made it possible to change the semantics of the function
to handle the M3-Source header missing, instead of skipping the
function.

### CODE REVIEW

File: src/query/api/v1/middleware/logging.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	"github.com/gorilla/mux"
+	"github.com/opentracing/opentracing-go"
+	"go.uber.org/zap"
+
+	"github.com/m3db/m3/src/query/util/logging"
+	"github.com/m3db/m3/src/x/instrument"
+)
+
+// ResponseLogging logs the response time if the request took longer than the configured threshold.
+func ResponseLogging(threshold time.Duration, iOpts instrument.Options) mux.MiddlewareFunc {
```
Comment (ryanhall07) [inline src/query/api/v1/middleware/logging.go:36]: this moved from the `util/logging` pkg.

File: src/query/api/v1/middleware/metrics.go [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+)
+
+var histogramTimerOptions = instrument.NewHistogramTimerOptions(
+	instrument.HistogramTimerOptions{
+		// Use sparse histogram timer buckets to not overload with latency metrics.
+		HistogramBuckets: instrument.SparseHistogramTimerHistogramBuckets(),
+	})
+
+// ResponseMetrics records metrics for the http response.
+func ResponseMetrics(iOpts instrument.Options) mux.MiddlewareFunc {
```
Comment (ryanhall07) [inline src/query/api/v1/middleware/metrics.go:42]: this was split out from the existing logging "middleware", into a separate middleware fn

File: src/query/api/v1/middleware/panic.go [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+	"github.com/m3db/m3/src/x/instrument"
+	xhttp "github.com/m3db/m3/src/x/net/http"
+)
+
+var highPriority = zap.LevelEnablerFunc(func(lvl zapcore.Level) bool {
+	return lvl >= zapcore.ErrorLevel
+})
+
+// Panic recovers from panics and logs the error/stacktrace if possible.
+func Panic(iOpts instrument.Options) mux.MiddlewareFunc {
```
Comment (ryanhall07) [inline src/query/api/v1/middleware/panic.go:42]: this was split out from the existing logging "middleware" into a separate fn

File: src/query/api/v1/middleware/tracing.go [hunk 4]
```diff
@@ -29,29 +29,17 @@ import (
...
-		CompressionMiddleware(),
-	}
-}
-
-// TracingMiddleware applies OpenTracing compatible middleware, which will start a span
+// Tracing applies OpenTracing compatible middleware, which will start a span
 // for each incoming request. Additionally if iOpts is non-nil the trace_id and span_id are added as fields to the
 // request scoped logger.
-func TracingMiddleware(tracer opentracing.Tracer, iOpts instrument.Options) mux.MiddlewareFunc {
+func Tracing(tracer opentracing.Tracer, iOpts instrument.Options) mux.MiddlewareFunc {
```
Comment (ryanhall07) [inline src/query/api/v1/middleware/tracing.go:42]: moved to a new file to be consistent with logging and metrics

File: src/query/util/queryhttp/queryhttp.go [hunk 5]
```diff
@@ -74,46 +63,19 @@ type RegisterOptions struct {
...
-	opts RegisterOptions,
-	middlewareOpts ...logging.MiddlewareOption,
-) error {
-	// Wrap requests with response time logging as well as panic recovery.
-	var (
-		route             *mux.Route
-		metrics           = newRouteMetrics(r.instrumentOpts)
-		middlewareOptions []logging.MiddlewareOption
-	)
-	postRequestOption := logging.WithPostRequestMiddleware(
```
Comment (ryanhall07) [inline src/query/util/queryhttp/queryhttp.go:90]: this "middleware" got decomposed into multiple functions

File: src/query/util/queryhttp/queryhttp.go [hunk 6]
```diff
@@ -74,46 +63,19 @@ type RegisterOptions struct {
...
-
-	wrapped := func(n http.Handler) http.Handler {
-		return logging.WithResponseTimeAndPanicErrorLogging(n, r.instrumentOpts,
-			middlewareOptions...)
+func (r *EndpointRegistry) Register(opts RegisterOptions) error {
+	if len(opts.Middleware) == 0 {
+		opts.Middleware = middleware.Default(r.instrumentOpts)
+	}
+	handler := opts.Handler
+	for i := len(opts.Middleware) - 1; i >= 0; i-- {
```
Comment (ryanhall07) [inline src/query/util/queryhttp/queryhttp.go:75]: important: installing the middleware

Comment (wesleyk) [reply src/query/util/queryhttp/queryhttp.go:75]: can we leave a comment why this ordering is important?

**Review by wesleyk — APPROVED** [review]
nice!

**Pull request merged**